Adaptors/MongoDB/src/TaskTable.cs

```diff
@@ -448,43 +454,69 @@ await foreach (var taskId in taskCollection.AsQueryable(sessionHandle)
+
+    Logger.LogDebug("Acquire task {task} on {podName}",
                     taskId,
-                    hostname);
-    var res = await taskCollection.UpdateManyAsync(x => x.TaskId == taskId && x.OwnerPodId == "",
-                                                   updateDefinition,
-                                                   cancellationToken: cancellationToken)
+                    ownerPodId);
+    var res = await taskCollection.FindOneAndUpdateAsync(filter,
+                                                         updateDefinition,
+                                                         new FindOneAndUpdateOptions<TaskData>
+                                                         {
+                                                           ReturnDocument = ReturnDocument.After,
+                                                         },
+                                                         cancellationToken)
                                   .ConfigureAwait(false);
 
-    switch (res.MatchedCount)
-    {
-      case 0:
-        var sessionHandle = sessionProvider_.Get();
-        var ownerPodId = await taskCollection.AsQueryable(sessionHandle)
-                            .Where(tdm => tdm.TaskId == taskId)
-                            .Select(model => model.OwnerPodId)
-                            .SingleAsync(cancellationToken)
-                            .ConfigureAwait(false);
-        Logger.LogInformation("Task {taskId} already acquired by {OtherOwnerPodId}", taskId, ownerPodId);
-        return false;
-      case 1:
-        return true;
-      default:
-        throw new ArmoniKException($"Error during acquisition of task {taskId}");
-    }
+    return res ?? await ReadTaskAsync(taskId,
+                                      cancellationToken)
+             .ConfigureAwait(false);
+  }
+
+  /// <inheritdoc />
+  public async Task<TaskData> ReleaseTask(string            taskId,
+                                          string            ownerPodId,
+                                          CancellationToken cancellationToken = default)
+  {
+    using var activity       = activitySource_.StartActivity($"{nameof(ReleaseTask)}");
+    var       taskCollection = taskCollectionProvider_.Get();
+
+    var updateDefinition = new UpdateDefinitionBuilder<TaskData>().Set(tdm => tdm.OwnerPodId,
+                                                                       "")
+                                                                  .Set(tdm => tdm.Status,
+                                                                       TaskStatus.Submitted);
+
+    var filter = new FilterDefinitionBuilder<TaskData>().Where(x => x.TaskId == taskId && x.OwnerPodId == ownerPodId &&
+                                                                    (x.Status == TaskStatus.Processing || x.Status == TaskStatus.Dispatched));
```

Comment:
`in` for mongodb request?

Reply:
I thought you could just change the part where you test for equality and still use the expression builder.